bpf/cgroup/bpf_cgroup_release.c

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright Authors of Tetragon */
+
+#include "vmlinux.h"
+#include "api.h"
+
+#include "hubble_msg.h"
+#include "bpf_cgroup.h"
+#include "bpf_events.h"
+#include "bpf_cgroup_events.h"
+
+char _license[] __attribute__((section(("license")), used)) = "GPL";
+#ifdef VMLINUX_KERNEL_VERSION
+int _version __attribute__((section(("version")), used)) =
+	VMLINUX_KERNEL_VERSION;
+#endif
+
+/* Ensure to remove tracked cgroups from bpf map */
+__attribute__((section(("raw_tracepoint/cgroup_release")), used)) int
+tg_tp_cgrp_release(struct bpf_raw_tracepoint_args *ctx)
+{
+	uint64_t cgrpid;
+	struct cgroup *cgrp;
+	struct cgroup_tracking_value *cgrp_track;
+
+	cgrp = (struct cgroup *)ctx->args[0];
+	cgrpid = get_cgroup_id(cgrp);
+	/* This should never happen unless our helper failed */
+	if (cgrpid == 0)
+		return 0;
+
+	cgrp_track = map_lookup_elem(&tg_cgrps_tracking_map, &cgrpid);
+	/* This was never tracked exit now */
+	if (!cgrp_track)
+		return 0;
```

Comment:
Hmm we don't track deep nested cgroups, we stop at current tetragon cgroup level or at max 32 int (arbitrary limit), so this could mean two things:
1. The cgroup is deep nested below are limit , could be case but how systemd organize hierarchies 32 is fine, and if unprivileged users keep creating nested cgroupv2 hierarchies we don't care we should attach them all the ancestors that we care about.

2. Maybe race conditions when we read already running process and cgroups that started before tetragon.

In following patches I have one patch that will push those even if we miss them during startup, as long as they do a fork() and execve() we catch and save them in the tracking cgroup map, however if they don't do the above and we miss between reading /proc and cgrooup_mkdir tracepoint being really active then yeh!

So yeh solution here check the level of the cgroup if it is not over our tracked limit then we did probably miss it, and should report this. Note added.